crates/bevy_ui/src/layout_components.rs

```diff
@@ -0,0 +1,561 @@
+//! Components used to control the layout of UI-[`Node`](super::Node) entities.
+use crate::{Size, UiRect, Val};
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::prelude::Component;
+use bevy_reflect::prelude::*;
+use serde::{Deserialize, Serialize};
+
+/// Grouping of core control parameter of the layout for this node.
+#[derive(Component, Copy, Clone, PartialEq, Debug, Serialize, Deserialize, Reflect)]
+#[reflect(PartialEq, Serialize, Deserialize)]
+pub struct LayoutControl {
```

Comment:
I also think theres a strong argument to put all layout related things under a single `Layout` component (aka dont break things up, just rename `Style` to `Layout`). Outside of the more granular change detection (which doesn't matter for the actual Taffy layout impl and will likely rarely come up in user code), I don't think breaking this up meaningfully improves UX. And it actively degrades it in a number of places:
* Introduces more "virtual" concept nesting users need to contend with: Spacing (not a css term), Flex Container/Item (a css term, but not one users directly interact with in flex apis), SizeConstraints (not a CSS term, `size_constraints.min` diverges from `min_size`), inset (a css term, but not one commonly used, and never as `layout.inset.top`)
* Discoverability is more challenging: users setting layouts now need to know how to access 5 components instead of 1. Additionally, having one component means you can discover all fields via autocomplete.
* Ergonomics downgrade: a user that wants to switch a node to "flex", set the flex container fields, and set the flex item fields now needs to write `Query<(&mut LayoutControl, &mut FlexContainer, &mut FlexItem)>` . That is a pretty big mouthful compared to `Query<&mut Layout>`.
    ```rust
    // centralized
    fn system(query: Query<&mut Layout>) {
      let mut layout = query.single();
      layout.display = Display::Flex;
      layout.flex_direction = Direction::Row;
      layout.grow = 0.5;
    }

    // broken up
    fn system(query: Query<(&mut LayoutControl, &mut FlexContainer, &mut FlexItem)>) {
      let (mut layout_control, mut flex_container, mut flex_item) = query.single();
      layout_control.display = Display::Flex;
      flex_container.direction = Direction::Row;
      flex_item.grow = 0.5;
    }
    ```
* Ergonomics downgrade: more nesting and verticality during instantiation:
    ```rust
    // centralized
    commands.spawn(NodeBundle {
        position: Position::Absolute,
        flex_direction: Direction::Row,
        grow: 0.5,
        ..default()
    })

    // broken up 
    commands.spawn(NodeBundle {
        control: LayoutControl {
            position: Position::Absolute,
            ..default()
        },
        flex_container: FlexContainer {
            direction: Direction::Row,
            ..default()
        },
        flex_item: FlexItem {
            grow: 0.5,
            ..default()
        },
        ..default()
    })
    ```
In theory, breaking things up into components could make things more modular. But in practice, Taffy is monolithic and a good portion of these concepts are deeply interrelated. And the centralized `Display` enum and "black box" nature of taffy from a user perspective prevents people from inserting new layout approaches into this meaningfully.

If I ask myself "will users be happy about these changes", in the current state I think the answer is likely "no". Maybe I'm wrong, but I'm not seeing the wins here.

Reply:
I'm not sure what others' motivations for this change are (I rather agree that it's not great from an ergonomics point of view, but one key benefit would be lower memory usage in the common case that not all styles are set. Taffy's `Style` struct is currently around 250 bytes. It's up to  300+ with CSS Grid support in Taffy 0.3, and is only likely to increase in future when add support for things like Morphorm layout and RTL text/flow-relative styles. Granted, we're probably only talking about 5-10mb here for realistic UI node counts of ~10000.

Reply:
Are we going to support other layout algorithms in Bevy, e.g. grids?
In that case I think it might make sense to keep them separated

Reply:
Yes, that's my ultimate hope.

Reply:
A bit late to the party here:

> In theory, breaking things up into components could make things more modular.

The original intent behind this was to break everything up so that things like `Margin`, `Padding`, `Display`, `Direction` would be top level components in order to get close to the "behaviour" of CSS. 

Then over time it morphed to what you see here which is due to multiple good arguments & various pushbacks. 

I do agree with closing this particular PR, but I keep having the feeling that the original idea would be worthwhile still but it will require a different execution.

Reply:
It seems to me that the intent of the proposal is to make `NodeBundle` into a flat namespace. But currently that would entail each style being it's own bevy_ecs `Component`, which is considered undesirable for perf reasons. Perhaps a possble alternative would be some way of mapping bundles to components that didn't just assume that each property of the bundle struct is it's own component, but allows a custom mapping.

Reply:
> It seems to me that the intent of the proposal is to make NodeBundle into a flat namespace. But currently that would entail each style being it's own bevy_ecs Component, which is considered undesirable for perf reasons.

Yeah, you sum it up very well here. The idea was to achieve increased ergonomics and more granular change detection but I'm starting to get the feeling that this is trying to solve the symptom instead of the root cause. 

Ergonomic UI in Bevy is IMO more likely to be achieved with finding a better way to structure data-flow and high level components instead of just changing the api surface of what is already there. In short: We need a working design :)